Workers can run and execute tasks

mhrivnak opened the pull request:
Thanks to ttereshc aka @goosemania who contributed some of this code.

There is a known issue that resource-reserving tasks still cannot run, but
there is a fix in the works that will follow this PR.

Fixes bugs in pulp.tasking
Rearranges pulp.tasking to put related code together
Adds two tasks (importer and publisher delete) in pulp.app.tasks
Fixes import order problems of django app vs. celery app

https://pulp.plan.io/issues/2440

re #2440
re #2400
re #2408

goosemania commented on app/pulp/app/tasks/__init__.py:
```diff
@@ -0,0 +1,2 @@
+from pulp.app.tasks import repository  # noqa
```
Is there any particular reason to import task modules here and no the tasks themselves? No to put all tasks in the same namespace?

mhrivnak replied:
This could go multiple ways, and I'm not tied to this one. Here's how I settled on this one to try: If we import every single task, each of which is just a function, into the same namespace, then the function names will have to contain additional namespace to avoid collision. For example, ``delete_repository``, ``delete_importer``, ``delete_*``.

Would we do ``from repository import delete_repository``? That already looks slightly silly.

That would mean having a function at ``pulp.app.tasks.repository.delete_repository``. Why not just name it ``pulp.app.tasks.repository.delete`` and import the whole ``repository`` module?

That said, it's just semantics and naming, with minor changes. I'm not sure it matters a lot in any case.

Also, while typing this I've realized that the tasks in this PR should be moved to ``importer`` and ``publisher`` modules.

What do you think?

goosemania replied:
That sounds really good :+1:

goosemania commented on tasking/pulp/tasking/celery_app.py:
```diff
@@ -69,6 +66,10 @@ def initialize_worker(sender, instance, **kwargs):
     """
+    from pulp.tasking.services.worker_watcher import delete_worker
+    from pulp.tasking.services import storage
+    from pulp.tasking.constants import TASKING_CONSTANTS
```
It is not necessary to import `constants` and `storage` inside callables. There should be no harm if they will be in their usual place because we import `pulp.tasking` anyway: `from pulp.tasking.celery_instance import celery`

goosemania commented on tasking/pulp/tasking/celery_app.py:
```diff
@@ -92,6 +93,8 @@ def get_resource_manager_lock(name):
     :type  name:   basestring
     """
+    from pulp.app.models.task import TaskLock, Worker
+    from pulp.tasking.constants import TASKING_CONSTANTS
```
It is not necessary to import `constants` inside callables.

mhrivnak replied:
Moved.

goosemania commented on tasking/pulp/tasking/services/scheduler.py:
```diff
@@ -9,15 +10,20 @@
+
+# This import is here so that Celery will find our application instance. It's important that other
+# pulp and django code not get used until after the celery app is instantiated and does its "fixup"
+# of django.
```
I do not think that order of the imports is important in scheduler.
To start celerybeat apart form scheduler you specify `--app=pulp.tasking.celery_app:celery` and it will be handled first. As for any other workers the first thing celery will do is import this app module and start worker. All the fixups will happen then. So at the time of scheduler import, everything will be ready and imports can go here in any order. If it was not the case you would not able to import models outside of callables here. That's my understanding.

mhrivnak replied:
That makes sense! I've changed the order to the usual convention and everything still works.

goosemania commented on app/pulp/app/tasks/repository.py:
```diff
@@ -0,0 +1,22 @@
+
+@shared_task(base=UserFacingTask)
+def delete_publisher(repo_name, publisher_name):
+    models.Publisher.objects.filter(name=publisher_name, repository__name=repo_name).delete()
```
Maybe it makes sense to remove the implementation of both of these tasks from pulp2 codebase in this 3.0-dev branch. 
And also we should not forget about cleanup by plugins which happened in these tasks in pulp 2.

mhrivnak replied:
Will do!

mhrivnak commented on tasking/pulp/tasking/services/scheduler.py:
```diff
@@ -9,15 +10,20 @@
 from django.utils import timezone
 from django.db import IntegrityError
 
+os.environ.setdefault("DJANGO_SETTINGS_MODULE", 'pulp.app.settings')
```
Removed.

bmbouter commented on app/pulp/app/tasks/publisher.py:
```diff
@@ -0,0 +1,17 @@
+    """
+    Delete a publisher.
+
+    :param repo_name:       the name of a repository
```
This should be converted to use the new Google style docstring

bmbouter commented on app/pulp/app/models/task.py:
```diff
@@ -35,6 +35,29 @@ class ReservedResource(Model):
+
+        This function also provides randomization for worker selection.
+
+        :raises Worker.DoesNotExist: If all workers have ReservedResource entries associated with them.
```
This should be converted to use the new Google style docstring

bmbouter commented on tasking/pulp/tasking/celery_app.py:
```diff
@@ -14,15 +14,14 @@
-
-# This import is here so that Celery will find our application instance
+# This import is here so that Celery will find our application instance. It's important that other
+# pulp and django code not get used until after the celery app is instantiated and does its "fixup"
```
s/pulp/Pulp/
s/django/Django/
s/celery/Celery/

bmbouter commented on tasking/pulp/tasking/services/worker_watcher.py:
```diff
@@ -100,3 +102,42 @@ def handle_worker_offline(event):
+
+    Any tasks associated with this worker are explicitly canceled.
+
+    :param name:            The name of the worker you wish to delete.
```
This should be converted to use the new Google style docstring

bmbouter replied:
This should be converted to use the new Google style docstring

bmbouter commented on server/usr/lib/systemd/system/pulp_celerybeat.service:
```diff
@@ -7,7 +7,7 @@ Wants=network-online.target
 User=apache
 WorkingDirectory=/var/run/pulp/
-ExecStart=/usr/bin/celery beat --app=pulp.tasking.celery_instance.celery --scheduler=pulp.tasking.scheduler.Scheduler
+ExecStart=/usr/bin/celery beat --app=pulp.tasking.celery_app:celery --scheduler=pulp.tasking.services.scheduler.Scheduler
```
👍

bmbouter commented on server/usr/lib/systemd/system/pulp_resource_manager.service:
```diff
@@ -7,7 +7,7 @@ Wants=network-online.target
 User=apache
 WorkingDirectory=/var/run/pulp/
-ExecStart=/usr/bin/celery worker -A pulp.tasking.celery_app -n resource_manager@%%h\
+ExecStart=/usr/bin/celery worker -A pulp.tasking.celery_app:celery -n resource_manager@%%h\
```
👍

bmbouter commented on tasking/pulp/tasking/celery_app.py:
```diff
@@ -14,15 +14,14 @@
 
-# This import is here so Celery will discover all tasks
-import pulp.tasking.registry  # noqa
+celery.autodiscover_tasks()
```
This should get another line of whitespace above it separating it from imports

bmbouter commented on tasking/pulp/tasking/services/scheduler.py:
```diff
@@ -10,14 +10,13 @@
 
 # The import below is not used in this module, but it needs to be kept here. This module is the
-# first and only Pulp module to be imported by celerybeat, and by importing pulp.server.logs, it
+# first and only Pulp module to be imported by celerybeat, and by importing pulp.app.logs, it
```
I suspect this comment and the import on L#21 is probably out of date and can likely be removed along with the [`pulp.app.logs`](https://github.com/pulp/pulp/blob/3.0-dev/app/pulp/app/logs.py).

bmbouter replied:
This is the only reference I see to that module altogether:

```
[bmbouter@localhost pulp]$ git grep pulp.app.logs
tasking/pulp/tasking/services/scheduler.py:# first and only Pulp module to be imported by celerybeat, and by importing pulp.app.logs, it
tasking/pulp/tasking/services/scheduler.py:import pulp.app.logs  # noqa
```

bmbouter replied:
I [commented](https://pulp.plan.io/issues/2275#note-11) about this on the logging PR so these comments can be ignored.

bmbouter commented on app/pulp/app/models/task.py:
```diff
@@ -35,6 +35,30 @@ class ReservedResource(Model):
+class WorkerManager(models.Manager):
+    def get_unreserved_worker(self):
+        """
+        Find an unreserved Worker
```
Can the string 'Worker' be updated with the sphinx class path so it links to the actual `Worker` module?

bmbouter replied:
This is the only reference I see to that module altogether:

```
[bmbouter@localhost pulp]$ git grep pulp.app.logs
tasking/pulp/tasking/services/scheduler.py:# first and only Pulp module to be imported by celerybeat, and by importing pulp.app.logs, it
tasking/pulp/tasking/services/scheduler.py:import pulp.app.logs  # noqa
```

bmbouter replied:
Something like this for example:  https://github.com/bmbouter/pulp/blob/e806ab8a11a10fd1bbd9c7475f3187f0c8157d43/app/pulp/app/serializers/base.py#L86-L86

bmbouter replied:
This shows some of the other options like `:attr:` for example

http://www.sphinx-doc.org/en/1.4.9/domains.html#directive-py:class

bmbouter commented:
Can all of these commits be squashed?

bmbouter commented:
Do you want to remove the 'Work in Progress' label and also edit the title to remove [WIP]?

mhrivnak commented:
Yes and Yes.

bmbouter commented on app/pulp/app/models/task.py:
```diff
@@ -35,6 +35,30 @@ class ReservedResource(Model):
+        """
+        Find an unreserved :class:`pulp.app.models.Worker`
+
+        Return the Worker instance that has no :class:`pulp.app.models.ReservedResource`
```
Can Worker receive the linking also.

Note that you can use a ~ to write the full class path, but have Sphinx only show the classname itself.

bmbouter commented on app/pulp/app/models/task.py:
```diff
@@ -35,6 +35,30 @@ class ReservedResource(Model):
+        Find an unreserved :class:`pulp.app.models.Worker`
+
+        Return the Worker instance that has no :class:`pulp.app.models.ReservedResource`
+        associated with it. If all workers have ReservedResource relationships a,
```
Can ReservedResource receive the docs linking also?

bmbouter commented on tasking/pulp/tasking/tasks.py:
```diff
@@ -54,10 +52,10 @@ def _queue_reserved_task(name, task_id, resource_id, inner_args, inner_kwargs):
     :param name:          The name of the task to be called
     :type name:           basestring
-    :param inner_task_id: The UUID to be set on the task being called. By providing
+    :param task_id:       The UUID to be set on the task being called. By providing
```
Oh you are fixing a consistency problem, but can the parameter itself be renamed from `task_id` -> `inner_task_id`. It isn't actually the task_id of the _queue reserve resource itself.

bmbouter commented:
Can the commit message be improved to include a summary of changes, a link to the issue, and a `re #` syntax so it will link to the issue. I think a `re` is right because my additional PR is likely also needed and will need to be associated also before it can go to MODIFIED.

mhrivnak commented:
Yes, I will squash, write a real log message, etc. before removing the WIP status.

bmbouter commented on app/pulp/app/models/task.py:
```diff
@@ -35,6 +35,30 @@ class ReservedResource(Model):
+        Find an unreserved :class:`~pulp.app.models.Worker`
+
+        Return the Worker instance that has no :class:`~pulp.app.models.ReservedResource`
+        associated with it. If all workers have ReservedResource relationships a,
```
I think the comma should be moved to be after the word 'relationships'.

bmbouter commented on app/pulp/app/models/task.py:
```diff
@@ -35,6 +35,30 @@ class ReservedResource(Model):
+class WorkerManager(models.Manager):
+    def get_unreserved_worker(self):
+        """
+        Find an unreserved :class:`~pulp.app.models.Worker`
```
Can the idea that randomness is used be introduced in this 1-line description?

bmbouter commented on app/pulp/app/models/task.py:
```diff
@@ -35,6 +35,30 @@ class ReservedResource(Model):
+        associated with it. If all workers have ReservedResource relationships a,
+        :class:`pulp.app.models.Worker.DoesNotExist` exception is raised.
+
+        This function also provides randomization for worker selection.
```
s/function/method/

bmbouter commented on app/pulp/app/models/task.py:
```diff
@@ -35,6 +35,30 @@ class ReservedResource(Model):
 
 
+class WorkerManager(models.Manager):
+    def get_unreserved_worker(self):
```
not a standard style request, but consider adding a line of whitespace between the class definition and the first method.

bmbouter commented on app/pulp/app/models/task.py:
```diff
@@ -35,6 +35,30 @@ class ReservedResource(Model):
+        :raises Worker.DoesNotExist: If all workers have ReservedResource entries associated with
+                                     them.
+
+        :returns:          The Worker instance that has no reserved_resource
```
s/reserved_resource/ReservedResource/
Also consider replacing with `A random Worker instance that has no associated ReservedReservation entries`.

bmbouter commented on app/pulp/app/models/task.py:
```diff
@@ -35,6 +35,31 @@ class ReservedResource(Model):
+        :raises Worker.DoesNotExist: If all workers have ReservedResource entries associated with
+                                     them.
+
+        :returns:          a randomly-selected Worker instance that has no ReservedResource
```
s/a/A/

bmbouter commented on tasking/pulp/tasking/celery_app.py:
```diff
@@ -69,6 +69,8 @@ def initialize_worker(sender, instance, **kwargs):
     :param kwargs:   Other params (unused)
     :type  kwargs:   dict
     """
+    from pulp.tasking.services.worker_watcher import delete_worker
```
I really want to avoid this, but I'm not sure how. Any ideas?

goosemania replied:
It is so ^ because `delete_worker` imports models and also it imports `cancel` from `util` which also imports models.

bmbouter replied:
Yeah I don't think we can avoid this easily at this time

mhrivnak replied:
I suggest we go with this for now, and can try to improve it later if we find a better approach.

goosemania replied:
Is it worth creating a refactor task in redmine? (I can do it if needed) Or we are going to look for a better approach only in case of some problems so it will be resolved in some real issue?

bmbouter replied:
I don't think it's significant enough to have an issue tracking it

bmbouter commented on tasking/pulp/tasking/celery_app.py:
```diff
@@ -92,6 +94,7 @@ def get_resource_manager_lock(name):
     :param name:   The hostname of the worker
     :type  name:   basestring
     """
+    from pulp.app.models.task import TaskLock, Worker
```
Can we do this runtime lookup using the Django machinery. Can we try this instead?

```
from django.db.models import get_app
pulp_app =  get_app('pulp_app')
```

Then you could use `pulp_app.models.TaskLock` and `pulp_app.models.Worker`

I haven't tested this but I think it would work.

bmbouter replied:
Another possible option would be to shorten an import at the module level to `from pulp import app as pulp_app` and refer to the modules in this method the same way as identified ^.

mhrivnak replied:
Interesting tidbit:

```
./manage.py:1: RemovedInDjango19Warning: django.db.models.get_app is deprecated.
```

goosemania replied:
I guess it is expected to use `app.get_model(app_label, model)`:
```
from django.apps import apps
Task = apps.get_model('pulp_app', 'Task')
```

mhrivnak replied:
I think we could do this:

```
from django.apps import apps  # yes you read that right

<other code here>

def get_resource_manager_lock(name):
    TaskLock = apps.get_app_config('pulp_app').models['tasklock']
```

This avoids doing an import, but I'm not sure it's an improvement. It requires knowing these two strings, which isn't very likable. The import seems simpler and more obvious to understand.

bmbouter replied:
I agree, this seems more fragile than just the 1 line import. I'm ok if there are no changes in this area.

bmbouter commented on tasking/pulp/tasking/services/worker_watcher.py:
```diff
@@ -100,3 +102,43 @@ def handle_worker_offline(event):
+
+def delete_worker(name, normal_shutdown=False):
+    """
+    Delete the :class:`~pulp.app.models.Worker` with name from the database, cancel any associated
```
I believe autodoc expects the short description to be 1 line. I relies on that because that is what the docstring pep requires. Can this be shortened to 1 line < 100 characters somehow?

bmbouter commented on tasking/pulp/tasking/services/worker_watcher.py:
```diff
@@ -100,3 +102,43 @@ def handle_worker_offline(event):
+    tasks and reservations
+
+    If the worker shutdown normally, no message is logged, otherwise an error level message is
+    logged. Default is to assume the worker did not shut down normally.
```
The second sentence is a fragment.

bmbouter commented on tasking/pulp/tasking/services/worker_watcher.py:
```diff
@@ -100,3 +102,43 @@ def handle_worker_offline(event):
+    Delete the :class:`~pulp.app.models.Worker` with name from the database, cancel any associated
+    tasks and reservations
+
+    If the worker shutdown normally, no message is logged, otherwise an error level message is
```
I think the comma after logged should either be a semicolon or split into two separate sentences.

bmbouter commented on tasking/pulp/tasking/services/worker_watcher.py:
```diff
@@ -100,3 +102,43 @@ def handle_worker_offline(event):
+    If the worker shutdown normally, no message is logged, otherwise an error level message is
+    logged. Default is to assume the worker did not shut down normally.
+
+    Any resource reservations associated with this worker are cleaned up by this function.
```
Does it make sense to s/resource reservations/ReservedResources/ ?

bmbouter commented on tasking/pulp/tasking/services/worker_watcher.py:
```diff
@@ -100,3 +102,43 @@ def handle_worker_offline(event):
+
+    Any resource reservations associated with this worker are cleaned up by this function.
+
+    Any tasks associated with this worker are explicitly canceled.
```
Does it make sense to s/tasks/Task/ to refer to the model more explicitly?

bmbouter commented on tasking/pulp/tasking/services/worker_watcher.py:
```diff
@@ -100,3 +102,43 @@ def handle_worker_offline(event):
+
+    Any tasks associated with this worker are explicitly canceled.
+
+    :param name:            The name of the worker you wish to delete.
```
s/you wish//

bmbouter commented on tasking/pulp/tasking/tasks.py:
```diff
@@ -1,23 +1,21 @@
 from pulp.exceptions import MissingResource, PulpException
-from pulp.tasking import storage
+from pulp.tasking import util
+from pulp.tasking.services import storage
```
This should be moved down so it comes after `pulp.tasking.celery_instance`

bmbouter commented on tasking/pulp/tasking/tasks.py:
```diff
@@ -40,7 +38,7 @@ class PulpTask(CeleryTask):
 
 @task(base=PulpTask, acks_late=True)
-def _queue_reserved_task(name, task_id, resource_id, inner_args, inner_kwargs):
+def _queue_reserved_task(name, inner_task_id, resource_id, inner_args, inner_kwargs):
```
thank you

bmbouter commented on tasking/pulp/tasking/tasks.py:
```diff
@@ -75,7 +73,7 @@ def _queue_reserved_task(name, task_id, resource_id, inner_args, inner_kwargs):
 
         try:
-            worker = _get_unreserved_worker()
+            worker = Worker.objects.get_unreserved_worker()
```
👍

bmbouter commented on tasking/pulp/tasking/util.py:
```diff
@@ -0,0 +1,56 @@
+
+def cancel(task_id):
+    """
+    Cancel the task that is represented by the given task_id. This method cancels only the task
```
Can a 1-line description be added above this. I think autodoc will do better with that.

bmbouter replied:
I know you didn't write this (I probably did) but anyway

bmbouter commented on tasking/pulp/tasking/util.py:
```diff
@@ -0,0 +1,56 @@
+def cancel(task_id):
+    """
+    Cancel the task that is represented by the given task_id. This method cancels only the task
+    with given task_id, not the spawned tasks. This also updates task's state to 'canceled'.
```
s/not the/not/

bmbouter commented on tasking/pulp/tasking/util.py:
```diff
@@ -0,0 +1,56 @@
+
+def get_current_task_id():
+    """"
+    Get the current task id from celery. If this is called outside of a running
```
Can a 1-line description be added above here. I know you didn't write this originally.

bmbouter commented on tasking/pulp/tasking/util.py:
```diff
@@ -0,0 +1,56 @@
+def get_current_task_id():
+    """"
+    Get the current task id from celery. If this is called outside of a running
+    celery task it will return None
```
s/celery/Celery/

bmbouter replied:
This sentence needs a period.

bmbouter commented on tasking/pulp/tasking/util.py:
```diff
@@ -0,0 +1,56 @@
+    Get the current task id from celery. If this is called outside of a running
+    celery task it will return None
+
+    :return: The ID of the currently running celery task or None if not in a task
```
s/ID/UUID/  ?

bmbouter replied:
s/celery/Celery/
s/None/None,/

mhrivnak commented:
Thanks for all the feedback. I fixed a ton of doc block items, but I'd like to leave the rest for another time since I didn't write them (I just moved them). My feeling is that fixing so many things about these docs that I didn't write is becoming a distraction and pulling my focus away from the work I did do with this PR.

For future review, I suggest you consider saving such requests for a follow-up effort. A small number of changes while in the neighborhood can be convenient and productive, but it can quickly overwhelm the original goals of the PR.

bmbouter commented:
Thanks @mhrivnak this looks good. I can imagine how a lot of comments can be overwhelming, so I'll think more about that. The problem with follow-up efforts of this size is that they rarely get done. Anyway, a lot of the docblock improvements did get done, so I appreciate that. LGTM. 🎆 🎈
